Galactic: for_each_callback_group backport

adityapande-1995 opened the pull request:
This PR aims takle https://github.com/ros-simulation/gazebo_ros_pkgs/issues/1296 by backporting https://github.com/ros2/rclcpp/pull/1723 to galactic without breaking ABI. To do this, I added a new non virtual function in node_base class, and added static members to it to keep track of a global map of mutexes. This map stores the mutexes for each node_base instance. However, while using the new ```for_each_callback_group``` method, the pointer of type ```NodeBaseInterface``` has to be typecasted to ```NodeBase```. The global map of mutexes is written as a separate class and access to it is protected by an internal mutex.

# TODO 
* General code cleanup, linting issues
* Modify ```static_executor_entities_collector```
* Modify the corresponding test cases

clalancette commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -70,6 +70,8 @@
 #include "rclcpp/timer.hpp"
 #include "rclcpp/visibility_control.hpp"
 
+typedef std::function<void (rclcpp::CallbackGroup::SharedPtr)> CallbackGroupFunction;
```
We should remove this, and instead rely on the version that lives in node_base.hpp.

clalancette commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -153,6 +155,9 @@ class Node : public std::enable_shared_from_this<Node>
   const std::vector<rclcpp::CallbackGroup::WeakPtr> &
   get_callback_groups() const;
 
+  RCLCPP_PUBLIC
+  void for_each_callback_group(const CallbackGroupFunction & func);
```
We may as well make this one as much like the one on the `master` branch as possible.  That means copying over the style and the docblock as well (the only thing that should be different in the end is the exact namespace that `CallbackGroupFunction` is in).

clalancette commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -1310,9 +1315,9 @@ class Node : public std::enable_shared_from_this<Node>
   const std::string sub_namespace_;
   const std::string effective_namespace_;
 };
-
 }  // namespace rclcpp
 
+
```
Both of these whitespace changes are unnecessary here.  Just leave them as-is.

clalancette commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -18,13 +18,26 @@
 #include <memory>
 #include <string>
 #include <vector>
+#include <unordered_map>
+#include <mutex>
```
Alphabetize these with the ones above.

clalancette commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -18,13 +18,26 @@
 #include <vector>
+#include <unordered_map>
+#include <mutex>
+#include <iostream>
```
Remove this one, as it isn't needed.

clalancette commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -18,13 +18,26 @@
 #include "rclcpp/node_interfaces/node_base_interface.hpp"
 #include "rclcpp/visibility_control.hpp"
 
+#if 1
+#define DPRINT(x) std::cout << "DPRINT: " << x << std::endl;
+#else
+#define DPRINT(x)
+#endif
```
Remove this debug code.

clalancette commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -18,13 +18,26 @@
+#define DPRINT(x)
+#endif
+
+typedef std::function<void (rclcpp::CallbackGroup::SharedPtr)> CallbackGroupFunction;
+
+class map_of_mutexes;
```
These should both be in the namespace.  I would suggest putting them into the `rclcpp::node_interfaces` namespace, but other namespaces under `rclcpp` would work as well.

Also, make sure to `#include <functional>` at the top of this file to get `std::function`.

clalancette commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -145,4 +164,22 @@ class NodeBase : public NodeBaseInterface
+    // Members
+    std::unordered_map<const rclcpp::node_interfaces::NodeBase* , std::shared_ptr<std::mutex> > data;
+    std::mutex internal_mutex;
+
```
Unnecessary blank space.

clalancette commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -145,4 +164,22 @@ class NodeBase : public NodeBaseInterface
+    void delete_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBase* nodebase);
+
+    // Members
+    std::unordered_map<const rclcpp::node_interfaces::NodeBase* , std::shared_ptr<std::mutex> > data;
```
```suggestion
    std::unordered_map<const rclcpp::node_interfaces::NodeBase*, std::shared_ptr<std::mutex>> data;
```

clalancette commented on rclcpp/src/rclcpp/executor.cpp:
```diff
@@ -268,21 +268,26 @@ Executor::add_node(rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_pt
   std::atomic_bool & has_executor = node_ptr->get_associated_with_executor_atomic();
   if (has_executor.exchange(true)) {
-    throw std::runtime_error("Node has already been added to an executor.");
+    throw std::runtime_error(
+            std::string("Node '") + node_ptr->get_fully_qualified_name() +
+            "' has already been added to an executor.");
```
This is an unrelated change, so let's not put it into this PR.

clalancette commented on rclcpp/src/rclcpp/executor.cpp:
```diff
@@ -190,14 +191,13 @@ Executor::add_callback_groups_from_nodes_associated_to_executor()
-      std::for_each(
-        group_ptrs.begin(), group_ptrs.end(),
-        [this, node](rclcpp::CallbackGroup::WeakPtr group_ptr)
+      auto node_base_type = std::dynamic_pointer_cast<rclcpp::node_interfaces::NodeBase>(node);
```
This is a bit of bikeshedding, but I feel like `node_base` is a better name for this variable.  This is the instantiation of a type, not the type itself.  Same goes a few times below.

clalancette commented on rclcpp/src/rclcpp/node_interfaces/node_base.cpp:
```diff
@@ -46,6 +46,20 @@ NodeBase::NodeBase(
+  if (this->count_of_instances == 0){
+    this->map_object_ptr = std::make_unique<map_of_mutexes>();
+    this->count_of_instances++;
+    DPRINT("NB constructor- Made a new mm object !");
```
Remove this debug print.

clalancette commented on rclcpp/src/rclcpp/node_interfaces/node_base.cpp:
```diff
@@ -46,6 +46,20 @@ NodeBase::NodeBase(
+    this->count_of_instances++;
+    DPRINT("NB constructor- Made a new mm object !");
+  }
+  else{
+    DPRINT("NB constructor- mm object already exists")
+  }
```
Remove this debug code.

clalancette commented on rclcpp/src/rclcpp/node_interfaces/node_base.cpp:
```diff
@@ -46,6 +46,20 @@ NodeBase::NodeBase(
 {
+
+  // Initialize map_of_mutex object if this is the first NodeBase instance
+  if (this->count_of_instances == 0){
```
```suggestion
  if (this->count_of_instances == 0) {
```

clalancette replied:
Also, it's not clear to me that this is actually a count.  That is, it is incremented here, but it is never decremented.  This feels more like it is a boolean flag of "has this been initialized yet?".  There is also the secondary problem of what happens if two nodes are created on two different threads and race here.  It might be better to rely on C++ static initialization for this object, though that has some of its own problems.

adityapande-1995 replied:
I'll change that to a bool and add appropriate locking around it.

clalancette commented on rclcpp/src/rclcpp/node_interfaces/node_base.cpp:
```diff
@@ -310,3 +327,64 @@ NodeBase::resolve_topic_or_service_name(
+//  ***************** Implementation of map_of_mutexes ***************
+map_of_mutexes::map_of_mutexes()
+{
+  DPRINT("mm constructor called")
```
Remove this DPRINT and all of the ones below.

clalancette reviewed (changes requested):
I left a preliminary round of review.  I would also suggest rebasing all of the individual commits into a single commit, adding a useful commit message, and signing it properly (which will make the DCO bot happy).

adityapande-1995 commented:
Changes:
* Squashed commits, signed 
* Removed debug prints, macro
* removed ``typedef`` lines, ``CallbackGroupFunction`` now works the same way as in thr olling commit
* Changed variable name to ``node_base``
* ``map_of_mutexes`` now in ``rclcpp::node_interfaces`` namespace
* ``count_of_instances`` changed to boolean flag, protected by a mutex now
* Added docstring for ``for_each_callback_group`` in ``node.cpp`` and ``lifecycle_node.cpp``
* Removed blank spaces

TODO
* Modify tests
* Make changes to ``static_entities_collector``
* general cleanup, linting fixes, ament, uncrustify checks

adityapande-1995 commented:
TODO :
- CI failures in ```test_node.cpp```
- Change test cases to use ``for_each_callback_group``

adityapande-1995 commented:
TODO: Modify ```test_allocator_memory_strategy.cpp``` test to use ```for_each_callback_group```

ivanpauno commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -70,6 +70,7 @@
 #include "rclcpp/timer.hpp"
 #include "rclcpp/visibility_control.hpp"
 
+
```
do we need this new line?

adityapande-1995 replied:
will delete

ivanpauno commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -15,27 +15,42 @@
 #define RCLCPP__NODE_INTERFACES__NODE_BASE_HPP_
 
+#include <atomic>
+#include <functional>
```
do we need to include this header?

adityapande-1995 replied:
will delete

ivanpauno commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -15,27 +15,42 @@
 public:
   RCLCPP_SMART_PTR_ALIASES_ONLY(NodeBase)
 
+  static std::unique_ptr<map_of_mutexes> map_object_ptr;
```
why an unique_ptr?
Can't this just be an `static map_of_mutexes map_object;`

adityapande-1995 replied:
Will do

adityapande-1995 replied:
I think that was because ```this->map_object = map_of_mutexes()``` was not allowed as the class member ```std::mutex internal_mutex``` is copyable. To make the map_of_mutex static member ptr to an object, I'll have to change the ```internal_mutex``` as a pointer so that the copy constructor is valid.

clalancette replied:
> I think that was because in `node_base` constructor, `this->map_object = map_of_mutexes()`

But we shouldn't need that at all.  Once we have `map_of_mutexes` in static storage, there is no need to do an assignment to another static within the class; just use the global that already exists.

ivanpauno commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -15,27 +15,42 @@
   RCLCPP_SMART_PTR_ALIASES_ONLY(NodeBase)
 
+  static std::unique_ptr<map_of_mutexes> map_object_ptr;
+  static bool map_init_flag;
+  static std::mutex map_init_flag_mutex;
```
do we need to lazely initialize the map?
If yes, you can use std::call_once() and std::once_flag instead of doing it manually.

ivanpauno replied:
I don't think we need to lazely initialize the map, it seems to me that it can be initialized before main as all other globals.

clalancette replied:
Yeah, I agree.  I think that will be better, and will also get rid of the need for the extra mutex while lazily initializing.

ivanpauno commented on rclcpp/src/rclcpp/node.cpp:
```diff
@@ -605,3 +605,10 @@ Node::get_node_options() const
+void Node::for_each_callback_group(
+  const node_interfaces::NodeBaseInterface::CallbackGroupFunction & func)
+{
+  auto node_base = std::dynamic_pointer_cast<rclcpp::node_interfaces::NodeBase>(node_base_);
```
this is pretty unsafe, we're not even checking if the resulting ptr is not null.
We don't even need to add a method to `NodeBaseInterface` though, we can just add a free function that works like a thread safe interface to iterate over `get_callback_groups()`.
Of course, that only works if all callers are using the new interface, but that's also a problem if we add a method instead of a free function.

adityapande-1995 replied:
Will add a check to make sure the pointer is not NULL.

I had originally planned for a free function, but I discussed with @clalancette  and he suggested we make a non-virtual method in ```NodeBase``` and typecast ```NodebaseInterface``` members and then use the ```for_each_callback_group``` to make the code is consistent between rolling and the backport. Effectively the free function would do the same, typecast NodebaseInterface to NodeBase and then implement the for_each routine.

clalancette replied:
> We don't even need to add a method to `NodeBaseInterface` though, we can just add a free function that works like a thread safe interface to iterate over `get_callback_groups()`.

Yeah, as @adityapande-1995 points out, I don't think it matters whether we make it a free function or a non-virtual member function.  Either case preserves ABI; making it a non-virtual member function is slightly nicer in that you get `this` for free (rather than having to pass it in).

ivanpauno replied:
yes, the only difference is that the free function alternative doesn't need dynamic casting, as it can receive a reference to the base class directly.

I don't think we're extending `NodeBaseInterface` in another place, so I don't mind much.

ivanpauno replied:
example:

```cpp
void
for_each_calback_group_thread_safe(NodeBaseInterface & base_interface, FunctionCgT fn) {
  auto & mutex = get_cg_mutex_for_base_interface(&base_interface);
  std::lock_guard guard{mutex};
  for (const auto & cg : base_interface->get_callback_groups()) {
    fn(cg);
  }
}
```

In that case no dynamic cast is needed because [get_callback_groups](https://github.com/ros2/rclcpp/blob/e6733bf2c6cd40c8d8d26ec451042247577f0901/rclcpp/include/rclcpp/node_interfaces/node_base_interface.hpp#L131) is already a virtual method.

In that case, we will need the `NodeBaseInterface` constructor to register the mutex though (I think now `NodeBase` is doing that).

clalancette replied:
Ah, I see.  Yeah, that could work more nicely, leveraging the `get_callback_groups` that has to stay here.  @adityapande-1995 this is definitely worth exploring, as getting rid of the `dynamic_cast` will be a lot safer.

adityapande-1995 replied:
Will do

adityapande-1995 replied:
Added free function, removed the non virtual one and dynamic pointer casts

ivanpauno reviewed (commented):
I have a bunch of questions/comments.

clalancette commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -145,4 +156,23 @@ class NodeBase : public NodeBaseInterface
 }  // namespace node_interfaces
 }  // namespace rclcpp
 
+// Class to hold the global map of mutexes
+class rclcpp::node_interfaces::map_of_mutexes
+{
+public:
+  map_of_mutexes();
+  ~map_of_mutexes();
+
+  // Methods need to be protected by internal mutex
+  void create_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+  std::shared_ptr<std::mutex>
+  get_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+  void delete_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+
+  // Members
+  std::unordered_map<const rclcpp::node_interfaces::NodeBaseInterface *,
+    std::shared_ptr<std::mutex>> data;
+  std::mutex internal_mutex;
+};
```
I'll suggest just moving this class up above the `NodeBase` definition.  That way you can get rid of the forward-reference to it, and also just define it in the proper namespace.

clalancette commented on rclcpp/src/rclcpp/node_interfaces/node_base.cpp:
```diff
@@ -310,3 +316,51 @@ NodeBase::resolve_topic_or_service_name(
+  NodeBaseInterface * node_base_interface, const NodeBaseInterface::CallbackGroupFunction & func)
+{
+  auto mutex_ptr = NodeBase::map_object.get_mutex_of_nodebase(node_base_interface);
+  if (mutex_ptr) {
```
When can this condition be false?  I think we need to do "something" if this is false, either printing out an error or throwing an exception (depending on how severe the situation is).

adityapande-1995 replied:
This is probably not necessary, as the mutex is registered when a new ```NodeBase``` is created and deleted only when the ```NodeBase``` object is destroyed, so this should always return a non-NULL pointer. I've added a ```RCLCPP_ERROR```, though I think this check can be removed.

clalancette replied:
I'd say let's remove the check, then.

clalancette commented on rclcpp/src/rclcpp/node_interfaces/node_base.cpp:
```diff
@@ -310,3 +316,51 @@ NodeBase::resolve_topic_or_service_name(
+
+rclcpp::node_interfaces::map_of_mutexes NodeBase::map_object = map_of_mutexes();
+
+rclcpp::node_interfaces::map_of_mutexes::map_of_mutexes()
+{
+}
```
Since this is the default constructor, we don't need to define this.  Just remove this (and the declaration in the class).

clalancette commented on rclcpp/src/rclcpp/node_interfaces/node_base.cpp:
```diff
@@ -310,3 +316,51 @@ NodeBase::resolve_topic_or_service_name(
+  this->data.erase(nodebase);
+}
+
+rclcpp::node_interfaces::map_of_mutexes::~map_of_mutexes()
+{
+}
```
Same here; just remove the default destructor.

ivanpauno replied:
+1
I would also always stick to the rule of zero or rule of five, and never to the rule of three or any other intermediate solution.
See https://en.cppreference.com/w/cpp/language/rule_of_three#the_rule_of_three_five_zero#rule_of_zero, https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#c21-if-you-define-or-delete-any-copy-move-or-destructor-function-define-or-delete-them-all.

clalancette commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -145,4 +156,23 @@ class NodeBase : public NodeBaseInterface
 }  // namespace rclcpp
 
+// Class to hold the global map of mutexes
+class rclcpp::node_interfaces::map_of_mutexes
```
Nit: this can be `final`.

clalancette commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -145,4 +156,23 @@ class NodeBase : public NodeBaseInterface
+  get_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+  void delete_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+
+  // Members
+  std::unordered_map<const rclcpp::node_interfaces::NodeBaseInterface *,
+    std::shared_ptr<std::mutex>> data;
+  std::mutex internal_mutex;
```
I think these internal members can be `private`, no?

clalancette reviewed (changes requested):
Overall, I think this is getting close.  I've left a few more things to think about.

clalancette commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -16,26 +16,52 @@
 #include "rclcpp/node_interfaces/node_base_interface.hpp"
 #include "rclcpp/visibility_control.hpp"
 
+
```
Nit: unnecessary whitespace change.

clalancette commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -16,26 +16,52 @@
+  void delete_mutex_of_nodebase(const NodeBaseInterface * nodebase);
+
+private:
+  // Members
```
Nit: no need for this comment, it is pretty obvious.

clalancette commented on rclcpp/include/rclcpp/node_interfaces/node_base.hpp:
```diff
@@ -16,26 +16,52 @@
+
+private:
+  // Members
+  std::unordered_map<const NodeBaseInterface *, std::shared_ptr<std::mutex>> data;
+  std::mutex internal_mutex;
```
Nit: all of our member variables end with `_`, so this should be:

```suggestion
  std::unordered_map<const NodeBaseInterface *, std::shared_ptr<std::mutex>> data_;
  std::mutex internal_mutex_;
```

(obviously you'll have to update the code as well)

clalancette reviewed (changes requested):
I have a bunch of nits inline.  Once those are all fixed, I think this is good to go from my perspective, though I'd like to get @ivanpauno 's final thoughts before we run CI on this.

clalancette reviewed (approved):
This looks good to me with three things:

1. Green CI.
2. Running `abi-compliance-checker` before and after this PR to make sure we aren't changing ABI.
3. An approval from @ivanpauno and @cottsay

adityapande-1995 commented:
CI Jobs:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15059)](http://ci.ros2.org/job/ci_linux/15059/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=9792)](http://ci.ros2.org/job/ci_linux-aarch64/9792/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12745)](http://ci.ros2.org/job/ci_osx/12745/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15226)](http://ci.ros2.org/job/ci_windows/15226/)

cottsay commented on rclcpp/src/rclcpp/node_interfaces/node_base.cpp:
```diff
@@ -46,6 +48,9 @@ NodeBase::NodeBase(
   notify_guard_condition_is_valid_(false)
 {
+  // Generate a mutex for this instance of NodeBase
+  this->map_object.create_mutex_of_nodebase(this);
```
Is it standard practice to refer to static members using the instance reference like this? It seems misleading to me.

adityapande-1995 replied:
Changed to use ```NodeBase::``` instead of ```this->```

cottsay commented on rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node.hpp:
```diff
@@ -89,6 +89,7 @@
 #include "rclcpp_lifecycle/transition.hpp"
 #include "rclcpp_lifecycle/visibility_control.h"
 
+
```
Any particular reason for this spurious newline?

adityapande-1995 replied:
Removed

cottsay reviewed (approved):
Looks good from an ABI compatibility perspective. I had a couple of other comments inline, but nothing critical.

adityapande-1995 commented:
CI Jobs:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15063)](http://ci.ros2.org/job/ci_linux/15063/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=9795)](http://ci.ros2.org/job/ci_linux-aarch64/9795/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12748)](http://ci.ros2.org/job/ci_osx/12748/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15229)](http://ci.ros2.org/job/ci_windows/15229/)

adityapande-1995 commented:
@cottsay How should we handle the unstable windows CI ? The failures are in ```rosbag2_transport``` which doesn't seem to be related to this PR.

cottsay commented:
> How should we handle the unstable windows CI ? The failures are in rosbag2_transport which don't seem to be related to this PR.

If you're confident that they aren't a regression caused by this PR, then I think it's OK to ignore the test failures. It might be worth triggering a baseline build to be sure, though.

adityapande-1995 commented:
CI on raw galactic:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15072)](http://ci.ros2.org/job/ci_linux/15072/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=9804)](http://ci.ros2.org/job/ci_linux-aarch64/9804/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12757)](http://ci.ros2.org/job/ci_osx/12757/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15239)](http://ci.ros2.org/job/ci_windows/15239/)

ivanpauno reviewed (approved):
LGTM!

ivanpauno commented:
> @cottsay How should we handle the unstable windows CI ? The failures are in rosbag2_transport which don't seem to be related to this PR.

I can't tell if those failures are related or not, +1 to compare with a baseline build.

adityapande-1995 commented:
> CI on galactic branch of rclcpp:
> 
> * Linux [![Build Status](https://camo.githubusercontent.com/48d4d921feba7eb9f9202025f97c28885fd91e7320d80ce97b7c6cc3a0ef1678/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f6c696e7578266275696c643d3135303732)](http://ci.ros2.org/job/ci_linux/15072/)
> * Linux-aarch64 [![Build Status](https://camo.githubusercontent.com/f40c26929cf0db1cdd17143246747d3898e293b82bd94df7aeb67c0493a52f7f/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f6c696e75782d61617263683634266275696c643d39383034)](http://ci.ros2.org/job/ci_linux-aarch64/9804/)
> * macOS [![Build Status](https://camo.githubusercontent.com/6c39c08e643440c101c5803f3ceb7e5f7ae981781ded34fe708c74a88af237a0/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f6f7378266275696c643d3132373537)](http://ci.ros2.org/job/ci_osx/12757/)
> * Windows [![Build Status](https://camo.githubusercontent.com/4cda6a7d2c315a86fd48bfa4cd9b39ec5e3ffd8bcc617f4f46c59d75b46df03c/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f77696e646f7773266275696c643d3135323339)](http://ci.ros2.org/job/ci_windows/15239/)

So this CI was run by changing the branch of rclcpp to ```galactic```, i.e without my changes and it produced the same error in Windows. @cottsay @ivanpauno

ivanpauno commented:
> So this CI was run by changing the branch of rclcpp to galactic, i.e without my changes and it produced the same error in Windows. @cottsay @ivanpauno

LGTM!

cottsay reviewed (approved):
Thanks for double-checking!

adityapande-1995 commented:
ABI check report

![image](https://user-images.githubusercontent.com/18661155/130137415-5cad539a-c773-40df-8794-d06befdec739.png)
